MetaMask/metamask-extension
# Set sentry autoTrackSessions default

## Explanation

1. AutoSessionTracking defaults to true and operates by sending a session packet to sentry. This session packet does not appear to be filtered out via our beforeSend or FilterEvents integration. To avoid sending a request before we have the state tree and can validate the users preferences, we initiate this to false. Later, in startSession and endSession we modify this option and start the session or end the session manually. We also provide a toggleSession method that uses current state to determine whether to start or stop the session.

2. In sentry-install we call toggleSession after the page loads and state is available, this handles initiating the session for a user who has opted into MetaMetrics. This script is ran in both the background and UI so it should be effective at starting the session in both places.

3. In the MetaMetricsController the session is manually started or stopped when the user opts in or out of MetaMetrics. This occurs in the setParticipateInMetaMetrics function which is exposed to the UI via the MetaMaskController.

4. In actions.ts, after successfully calling the setParticipateInMetaMetrics background method, we tell sentry to toggle the session details appropriately.

5. Types for the global Sentry object, and the new methods added as part of this effort were added to global.d.ts in the types folder.

6. Another minor change is changing the throwTestError method so that it takes an optional input for the message text so that the Dedupe filter of sentry doesn't catch it for testing purposes. Without debug enabled in the sentry options you are unaware that the message is being dropped by that handler. 

7. Added an INVERSE test case, asserting that requests are NOT sent to sentry when the flag is false. 



## Screenshots/Screencaps



### Before

https://github.com/MetaMask/metamask-extension/assets/4448075/e97d6d6d-b57c-46e1-8178-1f0700a34312




### After

https://github.com/MetaMask/metamask-extension/assets/4448075/ce4db857-ab0f-4a00-84fe-61b2b91e62ba



## Manual Testing Steps With Existing Wallet
1. `yarn build dist` -- dev mode does not have a sentry global object attached to it. If you want to use dev mode, you need to comment out the 'return undefined' line in setupSentry with the comment block that talks about the bug with sentry in dev mode. 
2. Go to settings -> security and privacy -> toggle participate in metametrics to false
3. in full screen mode, open devtools and inspect the network tab. 
4. refresh the page and look for requests to sentry. There should be ZERO on this branch and ONE on develop. 

The next steps only work on this branch:
1. toggle metrics back on in settings.
2. In the console. run `window.stateHooks.throwTestError()`
3. See that the error is sent to sentry. (will only work once with no input, pass a string to throwTestError to get another error if you want to. Sentry dedupes the errors). 
4. toggle metrics back off in settings.
5. in the console run `window.stateHooks.throwTestError('foo')`
6. See error is not sent to sentry. 

## Test requests sent during onboarding
1. `yarn build:test` (easier to get to a clean wallet)
2. reset extension in plugins 
3. open extension for first time. 
4. On the welcome screen, open console and refresh. 
5. On develop: request sent to sentry,  on this branch: no request. 

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
